File: day2/a2.kt
import java.io.File

fun main() =
    File("input")
        .readLines()
        .map { it.split(' ') }
        .map { (id, v) -> id to v.toInt() }
        .fold(Triple(0, 0, 0)) { (x, d, a), (id, v) ->
          when (id) {
            "forward" -> Triple(x + v, d + v * a, a)
            "up" -> Triple(x, d, a - v)
            "down" -> Triple(x, d, a + v)
            else -> error("invalid token $id")
          }
        }
        .let { (x, d) -> x * d }
        .let(::println)

File: day2/a2.py
d, x, aim = (0, 0, 0)
with open('input', 'r') as input:
    for line in input.readlines():
        match line.split(' '):
            case ['forward', v]:
                x += int(v)
                d += int(v) * aim
            case ['up', v]:
                aim -= int(v)
            case ['down', v]:
                aim += int(v)
print(d * x)